onnxscript/rewriter/pattern.py

```diff
@@ -1029,37 +1033,118 @@ def _match_node_output(self, pattern_value: NodeOutputPattern, value: ir.Value)
+                else:
+                    unbound_values.append(f"output_{j}")
+            elif isinstance(value_pattern, Constant):
+                raise NotImplementedError("Constant values as return-values not supported.")
+        if unbound_values:
+            self._match.fail(f"Error: Output values not found: {unbound_values}")
+            return None
+        return output_values
+
+    def _match_single_output_node(
         self,
         model: ir.Model,
         graph_or_function: ir.Graph | ir.Function,
         node: ir.Node,
-        verbose: int = 0,
     ) -> MatchResult:
         del model
         del graph_or_function
-        self._verbose = verbose
-        self._matched: dict[NodePattern, ir.Node] = {}
-        self._match: MatchResult = MatchResult()
 
         pattern = self.pattern
         match = self._match
-        if len(node.outputs) != pattern.num_outputs:
+
+        if not pattern.has_single_output_node:
             return match.fail(
-                f"Number of node outputs mismatch: expected {pattern.num_outputs}, got {len(node.outputs)}."
+                "Internal Error: SimplePatternMatcher should not be used for patterns with multiple output nodes."
             )
-        if pattern._output_node is None:
+
+        if not self._match_node(pattern.output_node, node):
+            return match
+
+        output_values = self._get_output_values()
+        if output_values is None:
+            return match
+        if not _valid_to_replace(match.nodes, output_values):
+            return match.fail("Matched nodes have other uses preventing replacement.")
+
+        if len(node.outputs) != pattern.num_outputs:
             return match.fail(
-                "Internal Error: SimplePatternMatcher should not be used for patterns with multiple output nodes."
+                f"Number of node outputs mismatch: expected {pattern.num_outputs}, got {len(node.outputs)}."
             )
 
-        if self._match_node(pattern._output_node, node):
-            if not _valid_to_replace(match.nodes):
-                return match.fail("Matched nodes have other uses preventing replacement.")
+        match.outputs.extend(output_values)
+        return match
+
+    def _multi_match(self, candidate: dict[NodePattern, ir.Node]) -> MatchResult:
+        match = self._match
+        for pattern_node, node in candidate.items():
+            if not self._match_node(pattern_node, node):
+                return match
+        output_values = self._get_output_values()
+        if output_values is None:
+            return match
 
-        match.outputs.extend(node.outputs)
+        if not _valid_to_replace(match.nodes, output_values):
+            return match.fail("Matched nodes have other uses preventing replacement.")
+
+        match.outputs.extend(output_values)
         return match
 
+    def match(
+        self,
+        model: ir.Model,
+        graph_or_function: ir.Graph | ir.Function,
```

Comment:
Just curious: why is both model and graph required as args? Do we need to take info from the model?

Reply:
Not sure whether we need both now. This may be a remnant leftover from earlier code due to (i) Reading and/or updating opset imports, when using Proto-based IR, and (b) Refactored methods coming out of the rewriter into a shared interface used by both pattern-matching implementations.

Reply:
Looking over the multiple implementations of this abstract method, I think we can eliminate both model and graph. While we use graph, I think we can get it from the node (in the new IR, unlike in proto). However, I think this does require some careful thinking, especially in dealing with control-flow sub-graphs. Maybe a candidate for a separate PR.